lib/src/license_detection/confidence.dart

```diff
@@ -0,0 +1,260 @@
+  }
+  final diffs = getDiffs(
+    unknownLicense.tokens,
+    knownLicense.tokens,
+    matchRange,
+  );
+
+  final range = diffRange(tokensNormalizedValue(knownLicense.tokens), diffs);
+  final valuationDiffs = diffs.skip(range.start).take(range.end - range.start);
+
+  // Make an initial check on the diffs to see if their are any
+  // unacceptable substitutions made.
+  verifyNoVersionChange(valuationDiffs, knownLicense.identifier);
+  verifyNoGplChange(valuationDiffs, knownLicense.identifier);
+
+  final distance = scoreDiffs(valuationDiffs);
+  final confidence = confidencePercentage(
+    knownLicense.tokens.length,
+    distance,
+  );
+
+  if (confidence < threshold) {
+    throw LicenseMismatchException(
+        'Confidence $confidence is lesser than threshold $threshold');
+  }
+
+  return LicenseMatch(
+    unknownLicense.tokens
+        .skip(matchRange.input.start)
+        .take(matchRange.input.end - matchRange.input.start)
+        .toList(),
+    confidence,
+    knownLicense,
+    diffs,
+    Range(range.start, range.end),
+  );
+}
+
+/// Calculates the confidence percentage as 1 subtracted by the ratio of
+/// Levenshtein word distance to the number of tokens in known license.
+@visibleForTesting
+double confidencePercentage(int knownLength, int distance) {
+  if (knownLength == 0) {
+    return 1.0;
+  }
+  final confidence = 1.0 - (distance / knownLength);
+
+  return confidence;
+}
+
+@visibleForTesting
+List<Diff> getDiffs(
+  List<Token> inputTokens,
+  List<Token> knownTokens,
+  MatchRange matchRange,
+) {
+  final dmp = DiffMatchPatch();
+
+  final unknownText = tokensNormalizedValue(inputTokens
+      .skip(matchRange.input.start)
+      .take(matchRange.input.end - matchRange.input.start));
+
+  final knownText = tokensNormalizedValue(
+    knownTokens.take(matchRange.source.end),
+  );
+
+  final diffs = dmp.diffMain(unknownText, knownText);
+
+  dmp.diffCleanupSemantic(diffs);
+
+  return List.unmodifiable(diffs);
+}
+
+/// Returns the range in [diffs] that best resembles the known license text.
+///
+/// The range provides diffs from which the unknown text could be trimmed down to
+/// produce best resemble known text. Essentially it tries to remove parts of
+/// text in unknown license which are not a part of known license without affecting the
+/// confidence negatively i.e any false-negatives are not discarded while some of the
+/// true-negatives are discarded.
+@visibleForTesting
+Range diffRange(String known, List<Diff> diffs) {
+  var foundStart = false;
+  var seen = '';
+  var end = 0;
+  var start = 0;
+
+  for (end = 0; end < diffs.length; end++) {
+    if (seen.length > 1 && seen.trimRight() == known) {
+      break;
+    }
+
+    if (diffs[end].operation == Operation.insert ||
+        diffs[end].operation == Operation.equal) {
+      if (!foundStart) {
+        start = end;
+        foundStart = true;
+      }
+
+      seen += diffs[end].text + ' ';
```

Comment:
The loop can only return when there is an exact match between `seen` and `known`, it is seen that if we have an exact match at some point, then the `diffs` appearing after this point are irrelevant as they are not associated with known text and therefore we can safely ignore them.

Reply:
For the sake of demonstration, let's assume that `diffs[end].text` will be `1`, `2`, `3`, ... `9`.
At one point `seen == '1 2 3 '`. If at the point `known == '1 2 4'`, we can break the for loop, because `seen` and `known` will never be equal, as `seen` is always appended to. If at that point `known == '1 2 3 4', we can continue the loop.

Reply:
So I'll so a substring match every like
`seen != known.substring(0,seen.length)` or is getting a substring costly in dart?

Reply:
`known.startsWith(seen)` is probably cheaper - assuming you don't add the extra space to seen.

Reply:
The extra space will be common to both the `seen` and `known`.